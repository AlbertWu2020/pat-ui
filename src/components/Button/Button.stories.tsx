import React from "react";
import { action } from "@storybook/addon-actions";
import Button, { ButtonType, ButtonSize } from "./Button";

export default {
  title: "Button",
  component: Button,
};

const buttonStyle: React.CSSProperties = {
  marginRight: "5px",
  marginTop: "5px",
};

export const DefaultButton = () => (
  <Button onClick={action("Default Button clicked")}>Default Button</Button>
);

export const DiffSizeButton = () => (
  <div>
    <Button
      style={buttonStyle}
      btnSize={ButtonSize.Small}
      onClick={action("Small Button clicked")}
    >
      Small Button
    </Button>
    <Button style={buttonStyle} onClick={action("Default Button clicked")}>
      Default Button
    </Button>
    <Button btnSize={ButtonSize.Large} onClick={action("Large Button clicked")}>
      Large Button
    </Button>
  </div>
);

export const DiffTypeButton = () => (
  <div>
    <Button
      style={buttonStyle}
      btnType={ButtonType.Primary}
      onClick={action("Primary Button clicked")}
    >
      Primary Button
    </Button>
    <Button
      style={buttonStyle}
      btnType={ButtonType.Secondary}
      onClick={action("Secondary Button clicked")}
    >
      Secondary Button
    </Button>
    <Button
      style={buttonStyle}
      btnType={ButtonType.Danger}
      onClick={action("Danger Button clicked")}
    >
      Danger Button
    </Button>
    <Button
      style={buttonStyle}
      btnType={ButtonType.Default}
      onClick={action("Default Button clicked")}
    >
      Default Button
    </Button>
    <Button
      style={buttonStyle}
      disabled
      btnType={ButtonType.Default}
      onClick={action("DisabledDefault Button clicked should not work")}
    >
      Disabled Default Button
    </Button>
    <Button
      style={buttonStyle}
      btnType={ButtonType.Link}
      onClick={action("Link Button clicked")}
    >
      Link Button
    </Button>
    <Button
      disabled
      btnType={ButtonType.Link}
      onClick={action("Disabled Link Button clicked should not work")}
    >
      Disabled Link Button
    </Button>
  </div>
);
