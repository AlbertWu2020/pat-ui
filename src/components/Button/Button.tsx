import React from 'react';
import {classNames} from '../../utils/classNames';

export enum ButtonSize {
  Large = 'lg',
  Small = 'sm',
}
export enum ButtonType {
  Primary = 'primary',
  Secondary = 'secondary',
  Danger = 'danger',
  Default = 'default',
  Link = 'link',
}
export interface IButtonProps {
  className?: string;
  btnSize?: ButtonSize;
  btnType?: ButtonType;
  children?: React.ReactNode;
  disabled?: boolean;
}

type NativeButtonProps = IButtonProps &
  React.ButtonHTMLAttributes<HTMLButtonElement>;
type NativeAchorButtonProps = IButtonProps &
  React.AnchorHTMLAttributes<HTMLAnchorElement>;
type PatButtonProps = NativeButtonProps | NativeAchorButtonProps;

const Button: React.FC<PatButtonProps> = (props) => {
  const {btnSize, btnType, children, disabled, className, ...rest} = props;
  let styleClasses = classNames('btn', {
    [`btn-${btnType}`]: true,
    [`btn-${btnSize}`]: !!btnSize,
    disabled: !!(disabled && btnType === ButtonType.Link),
  });
  if (className) {
    styleClasses += ' ' + className;
  }

  let btn;
  if (btnType !== ButtonType.Link) {
    btn = (
      <button
        className={styleClasses}
        disabled={disabled}
        {...(rest as NativeButtonProps)}
      >
        {props.children}
      </button>
    );
  } else {
    if (disabled) {
      rest.onClick = (e: React.MouseEvent) => {
        e.preventDefault();
      };
    }
    btn = (
      <a className={styleClasses} {...(rest as NativeAchorButtonProps)}>
        {props.children}
      </a>
    );
  }
  return btn;
};

Button.defaultProps = {
  btnType: ButtonType.Default,
  disabled: false,
};

export default Button;
