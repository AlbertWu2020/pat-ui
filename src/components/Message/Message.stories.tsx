import React from 'react';
import Message, {messageType, iconType} from './Message';

import { faCircleNotch, faTimes } from '@fortawesome/free-solid-svg-icons';
import { FontAwesomeIcon } from '@fortawesome/react-fontawesome';

export default {
    title: 'Message',
    component: Message,
};

const messageStyle: React.CSSProperties = {
    marginRight: '5px',
    marginTop: '5px',
};

export const DefaultMessage = () => (
   <Message msgType={messageType.Basic} >
    <div className="header">
      Changes in Service
    </div>
    <div className="content">
      <p>We just updated our privacy policy here to better service our customers. 
        We recommend reviewing the changes.</p>
    </div>
   </Message>
);

export const DiffTypeMessage = () => (
    <div>
      <h4>Basic Message: </h4>
        <Message
        style={messageStyle}
        msgType={messageType.Basic}
        >
        <div className="header">
          Changes in Service
        </div>
        <div className="content">
          <p>We just updated our privacy policy here to better service our customers. 
            We recommend reviewing the changes.</p>
        </div>
        </Message>

        <h4>List Message: </h4>

        <Message
        style={messageStyle}
        msgType={messageType.List}
        >
        <div className="header">
            New Site Features
        </div>
        <div className="content">
            <ul className="list">
            <li>You can now have cover images on blog pages</li>
            <li>Drafts will now auto-save while writing</li>
            </ul>
        </div>
        </Message>

        <h4>Icon Message: </h4>
    
        <Message
        style={messageStyle}
        msgType={messageType.Icon}
        iconType={iconType.Spinner}
        >
        <FontAwesomeIcon className="circle-icon" icon={faCircleNotch} />
        <div className="spinner-message">
        <div className="header">
            Just one second
        </div>
        <div className="content">
            <p>We're fetching that content for you.</p>
        </div>
        </div>
        </Message>

        <h4>Dismiss Message: </h4>

        <Message
        style={messageStyle}
        msgType={messageType.Dismiss}
        iconType={iconType.Remove}
        >
        <FontAwesomeIcon className="remove-icon" icon={faTimes}
        onClick={() => {
          const element = document.querySelector('.remove-message');
          return element?.remove();
        }}/>
        <div className="header">
        Welcome back!
        </div>
        <div className="content">
        <p>This is a special notification which you can dismiss if you're bored with it.</p>
        </div>
        </Message>
    </div>
  );