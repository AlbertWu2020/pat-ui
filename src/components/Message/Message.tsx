import React, { 
    FC, 
    HTMLAttributes } from 'react';
import { classNames } from '../../utils/classNames';

export enum messageType {
    Basic = 'bs',
    List = 'ls',
    Icon = 'ic',
    Dismiss = 'ds'
}

export enum iconType {
    Spinner = 'spinner',
    Remove = 'remove'
}

export interface IMessageProps {
    /** set class name */
    className?: string;
    /** set message type */
    msgType?: messageType;
    /** set icon type */
    iconType?: iconType;
}

type AllMessageProps = IMessageProps & HTMLAttributes<HTMLElement>;

export const Message: FC<AllMessageProps> = (props) => {
    const {className, msgType, iconType, ...rest} = props;
    
    let styleClasses = classNames('msg', {
        [`msg-${msgType}`]: true,
        [`msg-${iconType}`]: !!iconType
    });

    if (className) {
        styleClasses += ' ' + className;
    }

    let message = (
        <div>
            <div className={styleClasses} {...(rest as AllMessageProps)}></div>
        </div>
    );

    return message;
};

Message.defaultProps = {
    msgType: messageType.Basic
};

export default Message;
