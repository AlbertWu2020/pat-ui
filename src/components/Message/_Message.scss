.msg {
    position: relative;
    background-color: rgb(238, 237, 237);
    border-radius: $border-radius;
    border-width: $border-width;
    border: $btn-border-width solid gray;
}