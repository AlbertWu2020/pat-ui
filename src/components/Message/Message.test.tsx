import React from 'react';

import {render, fireEvent} from '@testing-library/react';
import Message, {messageType, iconType} from './Message';
import { ButtonType } from '../Button/Button';

describe('Message', () => {
    it('should match snapshot', () => {
        const {asFragment} = render(<Message> Snapshot Message </Message>);
        expect(asFragment()).toMatchSnapshot();
    });

    it('should render correct message based on different props', () => {
        // Basic message
        const msgBasicProps = {
            msgType: messageType.Basic,
            className: 'test'
        }
        const msgBasicWrapper = render (
            <Message {...msgBasicProps}>Basic Message</Message>
        );
        const msgBasicElement = msgBasicWrapper.queryByText('Basic Message') as HTMLElement;
        expect(msgBasicElement).toBeInTheDocument();
        expect(msgBasicElement).toHaveClass('msg msg-bs test');

        // List message

        // Icon message

        // Dismiss message
        const msgDismissProps = {
            msgType: messageType.Dismiss,
            onClick: jest.fn(),
            iconType: iconType.Remove
        }
        const msgDismissWrapper = render(
            <Message {...msgDismissProps}>Dismiss Message</Message>
        );
        const msgDismissElement = msgDismissWrapper.queryByText('Dismiss Message') as HTMLElement;
        expect(msgDismissElement).toBeInTheDocument();
        expect(msgDismissElement).toHaveClass('msg msg-ds msg-remove');
        expect(msgDismissProps.onClick).toHaveBeenCalledTimes(0);
        fireEvent.click(msgDismissElement);
        expect(msgDismissProps.onClick).toHaveBeenCalledTimes(1);
    });
});