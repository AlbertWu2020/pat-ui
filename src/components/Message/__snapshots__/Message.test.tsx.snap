// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Message should match snapshot 1`] = `
<DocumentFragment>
  <div>
    <div
      class="msg msg-bs"
    >
       Snapshot Message
    </div>
  </div>
</DocumentFragment>
`;

exports[`Messgae should match snapshot 1`] = `
<DocumentFragment>
  <div>
    <div
      class="msg msg-bs"
    >
       Snapshot Message
    </div>
  </div>
</DocumentFragment>
`;
